wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -357,705 +401,903 @@ The different scenarios and the expected outcomes are,
++------------+     verify     +--------------+
+|   Human    |   ---------->  |    Tests     |
++------------+                +--------------+
 
 ```
-// Helper function to initialize a BankAccount with default data
-fun createDefaultBankAccount(): BankAccount {
-    return BankAccount(1000.0, "user", "password")
+
+Tests should focus on verifying the behavior of the implementation code, while humans should be able to easily understand and verify the correctness of the test code itself.
+
+Let's use a Restaurant class as an example to explain how to structure test functionalities effectively. The Restaurant class manages orders, calculates bills, and applies discounts.
+
+```kotlin
+class Restaurant(private val menu: Map<String, Double>) {
+  var items: List<String> = listOf()
+  private var discount: Double = 0.0
+
+  fun placeOrder(orderItems: List<String>) {
+    items = orderItems
+    println("Order placed: $items")
+    val totalBill = calculateBill()
+    println("Total bill after discount: ${totalBill - (totalBill * discount)}")
+  }
+
+  fun applyDiscount(isMember: Boolean, code: String, quarter: YearQuarter) {
+    discount = when {
+      isMember && code == "SAVE10" -> 0.10
+      code == "SAVE20" && quarter == YearQuarter.QUARTER1 -> 0.20
+      code == "SUMMERSALE" && quarter == YearQuarter.QUARTER2 -> 0.15
+      else -> 0.0
+    }
+    println("Discount applied: ${discount * 100}%")
+  }
+
+  private fun calculateBill(): Double {
+    return items.sumOf { menu[it] ?: 0.0 }
+  }
+}
+
+enum class YearQuarter {
+    QUARTER1, QUARTER2, QUARTER3, QUARTER4
 }
 ```
 
-This helps to reduce redundancy and maintain consistency across multiple test cases by centralizing repetitive setup tasks.
+It's important to understand how to segment or split functionalities in your tests to maintain clarity and avoid confusion. Knowing when to use helper functions and `@Before` / `@After` annotations effectively, and when to keep logic within the test cases themselves, ensures that your tests are both clear and maintainable. Let’s explore these concepts with a Restaurant Ordering System example.
+
+## 1. When and How to Divide Responsibilities
 
-### 2. Setup and Teardown using `@Before` and `@After`
+### a. Helper Functions
 
-Using `@Before` and `@After` annotations ensures that common setup and cleanup tasks are automatically executed before and after each test case, maintaining a clean and consistent test environment.
+Helper Functions are valuable for reducing redundancy in tests. They encapsulate **non-behavioral tasks**, ensuring that the focus remains on testing the core logic.
 
-**Example:**
+**Helper Function:**
 
+```kotlin
+// Helper function to initialize a Restaurant with a menu
+fun createRestaurantWithMenu(): Restaurant {
```

Comment:
Hmm... I thought that since they were provided as parameters and wouldn't be used in the core logic, they would only serve as default values for the menu (trying to replicate the initial setup). 

Now I've adapted the approach of sending test file exemptions as a reference and modified the Restaurant example it to accept a file path to store a order log history and using a utility function to load a text file for each test case Would this fit this situation?

Reply:
I think the log file idea was really good--it's a much better demonstration of 'extra details.' The order does seem to be relevant since the tests are dealing with the concept of 'valid' vs. 'invalid' items (which directly ties to the orders).